### PR TITLE
Heap-alloc-free Model Nodes, Growable Model Containers

### PR DESCRIPTION
**Pull request by josephbirkner**
Sorry for the monster PR 👹 

Concerns the following issues:

* #18 (Resolved)
* #19 (Resolved)
* #20 (Resolved)

### CODE REVIEW

**Comment by johannes-wolf** [comment]
Can you push again? CI was not triggered for some reason…

File: include/simfil/model/arena.h [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+    ArrayArenaIterator begin() const { return ArrayArenaIterator(*this, 0); }
+    ArrayArenaIterator end() const { return ArrayArenaIterator(*this, static_cast<ArrayIndex>(heads_.size())); }
+
+    ArrayRange range(ArrayIndex const& array) {return ArrayRange(begin(array), end(array));}
+
+private:
+    // Represents a chunk of an array in the arena.
+    struct Chunk
+    {
+        size_t offset = 0;      // The starting offset of the chunk in the segmented_vector.
```
[the remark this answers is not in the document]
Comment (josephbirkner) [reply include/simfil/model/arena.h:257]: I wouldn't support concurrently pushing to the same array 🤔

**Comment by josephbirkner** [comment]
Some numbers from bulk performance testing from NDS MapViewer (512 tiles, 131523 Features) using perfect initialCapacities for arrays/objects:

* Model got 11-15% bigger
* Query performance seems to be unchanged. This is unexpected, I would have thought that the removed heap allocs have an effect. But I guess the `ModelPool::resolve()` call and the new indirection from ArrayIndex to head Chunk also has a cost 🤔

**Comment by josephbirkner** [comment]
Measurements on Ryzen 3950X, MSVC, Release:

Old:

```
25: Query typeof Recursive                         100             1     3.94711 s 
25:                                         39.5789 ms    39.5301 ms    39.6317 ms 
25:                                         258.395 us    220.559 us    308.536 us 
25:                                                                                
25: Query field id Recursive                       100             1      3.8076 s 
25:                                         38.0556 ms    36.6594 ms    39.2351 ms 
25:                                         6.57192 ms    5.71923 ms    7.20166 ms 
25:                                                                                
25: Query field id                                 100             1    37.9534 ms 
25:                                         378.488 us      377.4 us    380.332 us 
25:                                         7.07713 us    4.67622 us    11.3391 us 
```

Original changes (with removed std::any but thread-safe arena):


```
32: Query typeof Recursive                         100             1     6.47582 s 
32:                                         63.8166 ms    63.6749 ms    63.9651 ms 
32:                                         740.166 us    658.765 us    843.627 us 
32:                                                                                
32: Query field id Recursive                       100             1     5.25274 s 
32:                                         53.4607 ms    53.3432 ms     53.581 ms 
32:                                         605.426 us    536.374 us    704.528 us 
32:                                                                                
32: Query field id                                 100             1    71.2307 ms 
32:                                         710.177 us    708.715 us    712.326 us 
32:                                         8.89042 us    6.41094 us    12.0637 us
```

Arena without lock:


```
32: Query typeof Recursive                         100             1     6.08096 s 
32:                                         61.3788 ms    61.1225 ms    61.6286 ms 
32:                                         1.29318 ms    1.13103 ms    1.52453 ms 
32:                                                                                
32: Query field id Recursive                       100             1     5.13459 s 
32:                                          51.411 ms     51.257 ms    51.5661 ms 
32:                                           789.5 us     695.11 us    941.401 us 
32:                                                                                
32: Query field id                                 100             1     69.541 ms 
32:                                          696.16 us    694.862 us    698.055 us 
32:                                         7.90253 us    5.82411 us    11.0849 us 
```


Simfil std::function removed:

```
32: Query typeof Recursive                         100             1     5.89598 s 
32:                                         58.0094 ms    57.8116 ms    58.1888 ms 
32:                                         958.802 us    819.133 us    1.13486 ms 
32:                                                                                
32: Query field id Recursive                       100             1     4.66803 s 
32:                                         46.1795 ms    46.0535 ms    46.2941 ms 
32:                                         611.813 us    537.145 us    718.216 us 
32:                                                                                
32: Query field id                                 100             1    67.4354 ms 
32:                                         654.849 us    652.334 us    663.475 us 
32:                                         21.2339 us    6.80426 us    47.8982 us 
```

With ModelNode::iterate(...) (last commit):

```
32: Query typeof Recursive                         100             1     5.58397 s 
32:                                         55.4138 ms    55.2198 ms    55.6399 ms 
32:                                         1.06206 ms    858.929 us    1.56577 ms 
32:                                                                                
32: Query field id Recursive                       100             1      4.2874 s 
32:                                         43.0501 ms    42.9664 ms    43.1345 ms 
32:                                         429.327 us    391.772 us    475.624 us 
32:                                                                                
32: Query field id                                 100             1    59.5717 ms 
32:                                         591.349 us    589.768 us    593.664 us 
32:                                         9.64389 us    7.12844 us    13.1763 us 
```

IMO, next step could be to add `ARRAY_ARENA_NO_GROWTH`, where the ArrayRange has a direct reference to the head chunk and the data. This ArrayRange could then be used in the Node::iterate() impls. This will remove the overhead from the head/continuation chunk lookups for every at() call.

**Comment by josephbirkner** [comment]
With `arena::iterate()`:

```
32: Query typeof Recursive                         100             1     4.02004 s 
32:                                         38.5438 ms    38.4231 ms    38.6848 ms 
32:                                         661.456 us    543.953 us    822.811 us 
32:                                                                                
32: Query field id Recursive                       100             1     2.87265 s 
32:                                         28.6523 ms    28.5987 ms    28.6979 ms 
32:                                         251.096 us    213.424 us    304.459 us 
32:                                                                                
32: Query field id                                 100             1    53.2869 ms 
32:                                         527.767 us    523.889 us    537.298 us 
32:                                         28.7552 us     8.7424 us    50.7735 us 
```

**Comment by josephbirkner** [comment]
I created #38 and #39 to hold on to some ideas for further performance improvements.

**Review by johannes-wolf — APPROVED** [review]

**Pull request merged**